### ISSUE TITLE
Images should have a meaningful alternative text

### ISSUE THREAD

**Issue by alvaromontoro** [jackdomleo7/Checka11y.css#35]
### Describe the new a11y feature or project enhancement

Images should nave a descriptive alternative text. Images in which the alternative text is "image" or "graphic" or a " " (blank space), are not providing enough information for assistive technologies.

### Describe the solution you'd like

This one is tricky because it can be really language-dependant. The same word that works in English (image) may work in French (image) but not in Spanish (imagen) or German (bild). Also, images don't have `::before` and `::after` pseudo-elements to display the error.

The language barrier is difficult to skip, the solution could include some common words in different languages that would trigger the rule, but having all of them may be complex. Some of them (e.g., a blank space). As for the highlight, the project could include an inline image in SVG or base64 with a message through the `content` property.

Then the selector/rule could be something like this:

```css
image[alt=" "],
image[alt="image"i],
image[alt="graphic"i],
image[alt="photo"i],
image[alt="picture"i],
... {
  border: 4px solid red;
  content: url(path-to-error-image)
}
```

### Link(s)

- [Codepen with demo of the code working](https://codepen.io/alvaromontoro/pen/WNwqQgj)
- [WebAIM: Alternative text](https://webaim.org/techniques/alttext/)

**Comment by jackdomleo7** [comment]
Thank you for submitting this issue.

I'm reluctant to add this in anytime soon, for two reason. The first reason as you pointed out, it would only catch English written websites, so we'll have a very specific feature. The second is that the `alt` text can be blank and it still be accessible, for example for decorative images - [see my example here](https://talk-a11y-at-the-forefront-of-our-minds.jackdomleo.dev/examples/01-images/solution) (be sure to inspect the images) and [see the source code here](https://github.com/jackdomleo7/TALK-accessibility-at-the-forefront-of-our-minds/blob/master/examples/01-images/solution/index.html).

I think this either shouldn't be a feature or requires more thought. 🙂

**Comment by danspratling** [comment]
I disagree @jackdomleo7.

You make 2 points, so I'll address them separately.

1. Blank alt text is valid `alt=""` but an empty space is invalid `alt=" "`. This solution doesn't discourage empty alt text, but it does ensure that alt text is meaningful.

2. The project is already English language only, with all errors being written exclusively in English. There is definitely more which could be done to go towards i18n but for now, we shouldn't discourage features which might only work for certain languages, because the output is only ever English language anyway.

I actually like that the image gets replaced here with an error image - that makes it super clear which images are wrong for visibly abled users and why. Would love it if we could still create a test-based error too though so that this remains completely accessible, but right now this is better than the current solution even if it's not perfect.

**Comment by jackdomleo7** [comment]
My apologies, I missed the blank space!

I suppose we could ignore the i18n for now and I'd be happy with that. My concern was focusing on that many devs/organisatiosn code in English, but their website may be an alternate language. So, having the errors in English is ok, but as @alvaromontoro pointed out, looking for English words in the `alt` attribute could be problematic.

I'm happy for a PoC of this to be done, then we'll see about merging it as a feature. 🙂

**Comment by danspratling** [comment]
I think the question here is more _"is it better than before"_.

Validating these alt tags in English:
* Provides extra context for English speaking users
* Does not provide any less context or fewer features _than the current state of the product_ in other languages.

So I'd say go for it and expand later.

The only issue I can see is if another language uses "image" to mean an entirely different thing which then gets picked up by this.

**Comment by jackdomleo7** [comment]
I'm happy with this.

The ideal is `checka11y.css` doesn't get pushed to production, so the devs can ignore the error if they know the wrong "image" was detected. But something we could improve down the line witn l18n.

**Comment by danspratling** [comment]
This sounds like a good approach. It's not perfect but if Firefox doesn't allow textual conversions this this is the next best thing.

Do we know what happens in safari?

**Comment by alvaromontoro** [comment]
Safari works fine as long as you don't add the alt text through the content (the solution above doesn't)

So doing this: `content: url(path-to-error-image);` works in Chrome and Safari, but doing this `content: url(path-to-error-image) / "alt text";` works only on Chrome. So better without it.

**Comment by alvaromontoro** [comment]
I can pick up this one if you want to. But I'd need a couple of things:

1. A list of invalid alternative text. We could go with a logic close to the one that WebAIM Wave uses, and trigger the error when the alt:
    - Contains only space characters (alt=" "),
    - Is just a word from this list: "image", "graphic", "photo", "photograph", "drawing", "painting", "artwork", "logo", "bullet", "button", "arrow", "more", "spacer", "blank", "chart", "table", "diagram", "graph", or "*"
    - Is an image file name (e.g., a string finishing in the text ".png", ".gif", ".jpg", or ",jpeg")
    - Begins with "graphic of", "bullet", or "image of"
    - Ends with "image" or "graphic"

   The first two are an extended version of the ones initially suggested. The third one would be easy to implement too. I fear the last two could trigger some false positives, and maybe it's better to avoid them (but easy to implement too).

2. A description of the image that will replace the incorrect image. I can go with something simple (just a blank square with the words "WARNING SUSPICIOUS ALT TEXT" [as in this example](https://codepen.io/alvaromontoro/pen/WNwqQgj)), or could try to add a small warning logo too (e.g., a triangle with an exclamation mark inside). It would be good to know if you have some preference for size image, colors, etc.

Please, let me know.

**Comment by danspratling** [comment]
1. I'm happy for you to take that approach. That sounds super thorough already. We can always update it if we need to go further. I would also say include all 5 bullets. We can reduce these later if users report it as being too aggressive.
2. How about "⚠️ Error: Alt text must describe the contents of the image". Then it prompts the users on how to improve it too.

**Comment by alvaromontoro** [comment]
As the project uses SCSS/Sass, I could create a more generic mixing that generates the SVG based on some parameters (e.g., icon, color, line1, line2, line3). That way it could be extensible if needed for other rules.

**Comment by jackdomleo7** [comment]
Yeah! Let's wait for the SCSS branch first? I'm currently waiting for the rush of Hacktoberfesters to get their PRs in first. Then I'll update the SCSS branch with `master`, then I'll convert the recently added features to SCSS, then we'll merge it. Then by all means, I'd be happy for this solution!
